File: python/programmers/dfsbfs/p4.py
from collections import defaultdict


class Graph:
    def __init__(self, tickets):
        self.tickets = tickets
        self.info = defaultdict(list)
        self.answer = []

    def makegraph(self):
        for location in self.tickets:
            start = location[0]
            end = location[1]
            self.info[start].append(end)
            self.info[start].sort(reverse=True)

    def dfs(self, via):
        if self.info[via] is None:
            return self.answer

        while self.info[via]:
            out = self.info[via].pop()
            self.dfs(out)
            self.answer.append(out)


def solution(tickets):
    g = Graph(tickets)
    g.makegraph()
    g.dfs("ICN")
    answer = ["ICN"] + g.answer[::-1]
    return answer

File: java/programmers/src/dfsbfs/p4.java
package dfsbfs;

import java.util.*;
import java.util.zip.CheckedOutputStream;

public class p4 {

    static String[][] tickets;

    static HashMap<String, PriorityQueue> graph;


    static ArrayList ans;


    public static void dfs(String via) {
        if (graph.get(via).isEmpty()) {
            return;
        }

        while (!graph.get(via).isEmpty()) {
            String next = (String) graph.get(via).poll();
            dfs(next);
            ans.add(next);
        }


    }


    public static void makeGraph() {


        for (String[] elm : tickets) {
            String start = elm[0];
            String end = elm[1];
            if (graph.containsKey(start)) {
                graph.get(start).add(end);


            } else {
                PriorityQueue<String> info = new PriorityQueue<String>();
                info.add(end);


                graph.put(start, info);

            }

        }
        System.out.println(graph);


    }


    public static void main(String[] args) {

        tickets = new String[][]{{"ICN", "SFO" }, {"ICN", "ATL" }, {"SFO", "ATL" }, {"ATL", "ICN" }, {"ATL", "SFO" }};
        graph = new HashMap<>();
        makeGraph();
        System.out.println(graph);
        ans = new ArrayList();
        dfs("ICN");

        Collections.reverse(ans);
        String[] returnans = new String[ans.size()+1];

        returnans[0] = "ICN";
        returnans.
    }
}
